// dewjs/lib/Common/Cache.js
/**
 * @author     LiuZhaoHui <hi.liuzhaoxin@gmail.com>
 * @link       http://www.eatbean.com/nodejs
**/

"use strict";

module.exports = Cache;

function Cache ()
{
		this._sectionData = this._cacheData = {};
}

Cache.prototype = 
{
    constructor : Cache, 
    __name__ : "Cache", 
    
    _sectionName : "", 
    
    // 获取配置节键名
    get allSectionKeys ()
    {
        return Object.keys(this._cacheData);
    }, 
    
    // 获取所有键名
    get allKeys ()
    {
        return Object.keys(this._sectionData);
    }, 
    
    // 获取配置节数
    get sectionCount ()
    {
        return Object.keys(this._cacheData).length;
    }, 
    
    // 获取配置数
    get count ()
    {
        return Object.keys(this._sectionData).length;
    }, 
    
    // 获取配置节名
    get sectionName ()
    {
        return this._sectionName;
    }, 
    
    // 获取当前配置节数据
    get current ()
    {
        return this._sectionData;
    }, 
       
    // 添加配置节
    addSection : function (name, value)
    {
        if (name in this._cacheData)
        {
            this.useSection(name);
        }
        else
        {
            this.setSection(name, value);
        }
    }, 
    
    // 设置配置节
    setSection : function (name, value)
    {
        this._cacheData[name] = value;
        this.useSection(name);
    }, 
    
    // 是否包含配置节
    hasSection : function (name)
    {
        return name in this._cacheData;
    }, 
    
    // 使用配置节
    useSection : function (name)
    {
        if (!(name in this._cacheData))
        {
            throw new Error(name + " section value does not exist");
        }
        
        this._sectionName = name;
        this._sectionData = this._cacheData[name];
        
        return this;
    }, 
    
    // 移除配置节
    removeSection : function (name)
    {
        if (name in this._cacheData)
        {
            if (this._sectionData === this._cacheData[name])
            {
                this._sectionData = null;
            }
            
            delete this._cacheData[name];
        }
    }, 
    
    // 移除所有配置节
    removeAllSection : function ()
    {
        this._cacheData = {};
    }, 
    
    // 添加
    add : function (name, value)
    {
        !(name in this._sectionData) && this.set(name, value);
    }, 
    
    // 设置
    set : function (name, value)
    {
        this._sectionData[name] = value;
    }, 
    
    // 是否包含
    has : function (name)
    {
        return name in this._sectionData;
    }, 
    
    // 获取
    get : function (name)
    {
        return this._sectionData[name];
    }, 
    
    // 移除
    remove : function (name)
    {
        if (name in this._sectionData)
        {
            delete this._sectionData[name];
        }
    }, 
    
    // 移除所有
    removeAll : function ()
    {
        this._sectionData = {};
    }
};
