// dewjs/lib/DataBase/MySQL/Connect.js
/**
 * @author     LiuZhaoHui <hi.liuzhaoxin@gmail.com>
 * @link       http://www.eatbean.com/nodejs
**/

"use strict";

var net = require("net");
var MysqlNative = require("./mysql-native/lib/mysql-native/socketclient.js");

var Connect = 
{
    __name__ : "Connect", 
    
    initialize : function (application)
    {
        
    }, 
    
    // 返回一个 Connect 实例
    getInstance : function (param)
    {
        var self = Object.create(this);
        self._id = param.id;
        self._host = param.host;
        self._port = param.port;
        self._username = param.username;
        self._password = param.password;
        self._dbname = param.dbname;
        self._charset = param.charset;
        
        return self;
    }, 
    
    create : function ()
    {
        var socket = new net.createConnection(this._port, this._host);
        socket.on("connect", this.connect_.bind(this, socket));
        socket.on("error", this.error_.bind(this, socket));
        socket.on("close", this.close_.bind(this));
        
        var mysql = new MysqlNative(socket);
        mysql.set("row_as_hash", false);
        mysql.auth(this._dbname, this._username, this._password);
        mysql.query("set names " + this._charset);
        
        return mysql;
    }, 
    
    connect_ : function (socket)
    {
        console.log("Mysql: \"" + this._id + "\" client listening at " 
        + socket.address().address + ":" + socket.address().port);
    }, 
    
    error_ : function (socket, exception)
    {
        console.log("Mysql: " + JSON.stringify(exception));
        socket.end();
    }, 
    
    close_ : function ()
    {
        // ...
    }
};

module.exports = Connect;
