// dewjs/lib/DataBase/MySQL/PersistentQuery.js
/**
 * @author     LiuZhaoHui <hi.liuzhaoxin@gmail.com>
 * @link       http://www.eatbean.com/nodejs
**/

"use strict";

var PersistentQuery = 
{
    __name__ : "PersistentQuery", 
    
    initialize : function ()
    {
        
    }, 
    
    // 返回一个 PersistentQuery 实例
    getInstance : function (query)
    {
        var self = Object.create(this);
        self._query = query;
        self._queue = [];
        
        return self;
    }, 
    
    openPersistentQuery : function ()
    {
        this._query._queue.push({type : 1, persistentQuery : this});
        this._query.execute();
        
        return this;
    }, 
    
    closePersistentQuery : function ()
    {
        var obj = this._object;
        obj.isFinish = true;
        obj.isPersistent = false;
        obj.query = this._query;
        obj.pools.returnObject(obj);
        
        return this._query;
    }, 
    
    addQuery : function (sql, callback, extInfo)
    {
        this._queue.push({type : 1, sql : sql, callback : callback, extInfo : extInfo});
        "_object" in this && this.execute();
    }, 
    
    execute : function ()
    {
        var obj = this._object;
        if (obj.isFinish === false) return;
        
        var item = this._queue.shift();
        if (item === undefined) return;
        
        obj.isFinish = false;
        obj.onfinish = item.callback;
        obj.mysql
          .query(item.sql)
          .on("field", obj.field)
          .on("row", obj.row)
          .on("end", item.extInfo ? obj.getExtInfo : obj.end)
          .on("error", obj.error);
    }
};

module.exports = PersistentQuery;
