// dewjs/lib/DataBase/Memcached/Connect.js
/**
 * @author     LiuZhaoHui <hi.liuzhaoxin@gmail.com>
 * @link       http://www.eatbean.com/nodejs
**/

"use strict";

var net = require("net");

var Connect = 
{
    __name__ : "Connect", 
    
    initialize : function (application)
    {
        
    }, 
    
    // 返回一个 Connect 实例
    getInstance : function (param)
    {
        var self = Object.create(this);
        self._id = param.id;
        self._host = param.host;
        self._port = param.port;
        
        return self;
    }, 
    
    create : function ()
    {
        var socket = new net.createConnection(this._port, this._host);
        socket.on("connect", this.connect_.bind(this, socket));
        socket.on("error", this.error_.bind(this, socket));
        socket.on("close", this.close_.bind(this));
        
        return socket;
    }, 
    
    connect_ : function (socket)
    {
        console.log("Memcached: \"" + this._id + "\" client listening at " 
        + socket.address().address + ":" + socket.address().port);
    }, 
    
    error_ : function (socket, exception)
    {
        console.log("Memcached: " + JSON.stringify(exception));
        socket.end();
    }, 
    
    close_ : function ()
    {
        // ...
    }
};

module.exports = Connect;
