// dewjs/lib/DataBase/Memcached/PersistentQuery.js
/**
 * @author     LiuZhaoHui <hi.liuzhaoxin@gmail.com>
 * @link       http://www.eatbean.com/nodejs
**/

"use strict";

var PersistentQuery = 
{
    __name__ : "PersistentQuery", 
    
    initialize : function ()
    {
        
    }, 
    
    // 返回一个 PersistentQuery 实例
    getInstance : function (query)
    {
        var self = Object.create(this);
        self._query = query;
        self._queue = [];
        
        return self;
    }, 
    
    openPersistentQuery : function ()
    {
        this._query._queue.push({type : 1, persistentQuery : this});
        this._query.execute();
        
        return this;
    }, 
    
    closePersistentQuery : function ()
    {
        var obj = this._object;
        obj.isFinish = true;
        obj.isPersistent = false;
        obj.query = this._query;
        obj.pools.returnObject(this._object);
        
        return this._query;
    }, 
    
    addQuery : function (cmd, callback)
    {
        this._queue.push({command : cmd, callback : callback});
        "_object" in this && this.execute();
    }, 
    
    execute : function ()
    {
        var obj = this._object;
        if (obj.isFinish === false) return;
        
        var item = this._queue.shift();
        if (item === undefined) return;
        
        obj.isFinish = false;
        obj.parser.onfinish = item.callback;
        obj.socket.write(item.command);
    }
};

module.exports = PersistentQuery;
