// dewjs/lib/Tool/Format.js
/**
 * @author     LiuZhaoHui <hi.liuzhaoxin@gmail.com>
 * @link       http://www.eatbean.com/nodejs
**/

"use strict";

var Format = 
{
    __name__ : "Format", 
    
    UNIT_TABLE : 
    {
        G : 1073741824, 
        M : 1048576, 
        K : 1024
    }, 
    
    formatBytes : function (bytes)
    {
        var T = this.UNIT_TABLE;
        if (bytes < T.K)
        {
            return bytes + "B";
        }
        else if (bytes < T.M)
        {
            return (bytes / T.K).toFixed(2) + "KB";
        }
        else if (bytes < T.G)
        {
            return (bytes / T.M).toFixed(2) + "MB";
        }
        else
        {
            return (bytes / T.G).toFixed(2) + "GB";
        }
    }, 
    
    ellipsis : function(value, len)
    {
        if(value.length > len)
        {
            return value.slice(0, len - 3) + "...";
        }
        
        return value;
    }, 
    
    // 把数字转换成美圆字符串
    usMoney : function (v)
    {
        v = (Math.round((v - 0) * 100)) / 100;
        v = (v == Math.floor(v)) ? v + ".00" : ((v * 10 == Math.floor(v * 10)) ? v + "0" : v);
        
        var ps = v.split('.');
        var whole = ps[0];
        var sub = ps[1] ? '.' + ps[1] : '.00';
        var r = /(\d+)(\d{3})/;
        
        while (r.test(whole))
        {
            whole = whole.replace(r, '$1' + ',' + '$2');
        }
        
        v = whole + sub;
        
        if (v.charAt(0) == '-')
        {
            return '-$' + v.substr(1);
        }
        
        return "$" + v;
    }
};

module.exports = Format;
