// dewjs/lib/Server/Cache/Protocol/Append.js
/**
 * @author     LiuZhaoHui <hi.liuzhaoxin@gmail.com>
 * @link       http://www.eatbean.com/nodejs
**/

"use strict";

module.exports = Append;

var Data = Dew("use", "Server.Cache.Data");

function Append (socket)
{
    this._socket = socket;
}

Append.prototype = 
{
    constructor : Append, 
    
    _mark : 0, 
    
    _value : "", 
    _valueLength : 0, 
    
    _state : 0, 
    STATE : 
    {
        START : 0, 
        FILED_START : 1, 
        FILED : 2, 
        DATA_START : 3, 
        DATA : 4, 
        END : 5, 
        ERROR : 6
    }, 
    
    RETURN_STATE : 
    {
        ERROR : -1, 
        PART : 0, 
        COMPLETE : 1
    }, 
    
    parse : function (chunk, i, len)
    {
        var c, index = i, S = this.STATE;
        
        for (; i < len; i++)
        {
            c = chunk[i];
            
            switch (this._state)
            {
                case S.START : 
                  // space
                  if (c === 0x20)
                  {
                      break;
                  }
                  
                  this._state = S.FILED_START;
                  
                case S.FILED_START : 
                  // space
                  if (c === 0x20)
                  {
                      break;
                  }
                  
                  index = i;
                  this._field = "";
                  this._state = S.FILED;
                  
                case S.FILED : 
                  // \r or \n
                  if (c === 0x0D || c === 0x0A)
                  {
                      this._state = S.DATA_START;
                      
                      if (index === i)
                      {
                          i--;
                          break;
                      }
                  }
                  // space
                  else if (c === 0x20)
                  {
                      this._state = S.FILED_START;
                  }
                  else
                  {
                      break;
                  }
                  
                  // <command name> <key> <flags> <exptime> <bytes> [noreply]\r\n
                  
                  this._field += chunk.toString("utf8", index, i);
                  
                  // <key>
                  if (this._mark === 0)
                  {
                      this._key = this._field;
                  }
                  // <flags>
                  else if (this._mark === 1)
                  {
                      this._flags = this._field | 0;
                  }
                  // <exptime>
                  else if (this._mark === 2)
                  {
                      this._exptime = this._field | 0;
                  }
                  // <bytes>
                  else if (this._mark === 3)
                  {
                      this._bytes = this._field | 0;
                  }
                  // [noreply]
                  else if (this._mark === 4)
                  {
                      if ("noreply" === this._field)
                      {
                          this._isNoreply = true;
                      }
                  }
                  else
                  {
                      i--;
                      this._state = S.ERROR;
                      break;
                  }
                  
                  // \n
                  if (c === 0x0A)
                  {
                      i--;
                  }
                  else
                  {
                      this._mark++;
                  }
                  
                  break;
                  
                case S.DATA_START : 
                  this._state = S.DATA;
                  
                  // \n
                  if (c === 0x0A)
                  {
                      index = i + 1;
                      break;
                  }
                  
                  index = i;
                  
                case S.DATA : 
                  // <data block>\r\n
                  if (this._valueLength < this._bytes)
                  {
                      this._valueLength++;
                  }
                  else if (this._valueLength === this._bytes && (c === 0x0D || c === 0x0A))
                  {
                      this._value += chunk.toString("utf8", index, i--);
                      this._state = S.END;
                  }
                  else
                  {
                      i--;
                      this._state = S.ERROR;
                  }
                  break;
                  
                case S.END : 
                  c = Data.append(this._key, this._value, this._flags, this._exptime, this._bytes);
                  
                  if (!("_isNoreply" in this))
                  {
                      if (c === Data.STATE.STORED)
                      {
                          this._socket.write("STORED\r\n");
                      }
                      else if (c === Data.STATE.NOT_STORED)
                      {
                          this._socket.write("NOT_STORED\r\n");
                      }
                  }
                  
                  this.index = i;
                  return this.RETURN_STATE.COMPLETE;
                  
                case S.ERROR : 
                  this.index = i;
                  this._socket.write("CLIENT_ERROR bad data chunk\r\n");
                  return this.RETURN_STATE.ERROR;
            }
        }
        
        if (this._state === S.DATA)
        {
            this._value += chunk.toString("utf8", index);
        }
        else if (this._state === S.FILED)
        {
            this._field += chunk.toString("utf8", index);
        }
        
        return this.RETURN_STATE.PART;
    }
};
