// dewjs/lib/Server/Web/Request/Form/PostedFile.js
/**
 * @author     LiuZhaoHui <hi.liuzhaoxin@gmail.com>
 * @link       http://www.eatbean.com/nodejs
**/

"use strict";

var fs = require("fs");
var path = require("path");

var Uuid = Dew("use", "Tool.Uuid");

var PostedFile = 
{
    __name__ : "PostedFile", 
    
    _contentType : "", 
    _contentLength : 0, 
    //_isSave : false, 
    
    initialize : function (HttpForm)
    {
        this._uploadTmpDir = HttpForm._uploadTmpDir;
    }, 
    
    // 返回一个 PostedFile 实例
    getInstance : function (multipart, filename)
    {
        var self = Object.create(this);
        self._fileName = path.basename(filename);
        self._filePath = path.normalize(this._uploadTmpDir + "/" + Uuid() + path.extname(self._fileName));
        
        //self._multipart = multipart;
        //self._requestContext = multipart._requestContext;
        self._writeStream = new fs.WriteStream(self._filePath);
        //self._resume = this.resume_.bind(self);
        
        return self;
    }, 
    
    write_ : function (buf, len)
    {
        this._contentLength += len;
        this._writeStream.write(buf/*, this._resume*/);
    }, 
    
    /*
    resume_ : function ()
    {
        ++this._multipart._writeCount && this._requestContext.resume();
    }, 
    */
    
    end_ : function ()
    {
        this._writeStream.end();
        //this._multipart = null;
        this._writeStream = null;
        //this._requestContext = null;
    }, 
    
    destroy_ : function ()
    {
        /*!this._isSave &&*/ fs.unlink(this._filePath);
    }, 
    
    // 获取上载文件的大小（以字节为单位）
    get contentLength ()
    {
        return this._contentLength;
    }, 
    
    // 获取客户端发送的文件的 MIME 内容类型
    get contentType ()
    {
        return this._contentType;
    }, 
    
    // 获取客户端上的文件的完全限定名称
    get fileName ()
    {
        return this._fileName;
    }, 
    
    // 获取上传后临时文件路径
    get filePath ()
    {
        return this._filePath;
    }, 
    
    // 保存上载文件的内容
    // 目前不支持跨磁盘保存文件
    saveAs : function (newFilePath, callback)
    {
        //this._isSave = true;
        fs.rename(this._filePath, newFilePath, callback);
    }
};

module.exports = PostedFile;
